streams/src/main/java/org/apache/kafka/streams/kstream/internals/suppress/KTableSuppressProcessor.java

```diff
@@ -17,70 +17,143 @@
+                                   final FullChangeSerde<V> valueSerde) {
+        requireNonNull(suppress);
         this.keySerde = keySerde;
         this.valueSerde = valueSerde;
+        maxRecords = suppress.getBufferConfig().maxRecords();
+        maxBytes = suppress.getBufferConfig().maxBytes();
+        suppressDurationMillis = suppress.getTimeToWaitForMoreEvents().toMillis();
+        buffer = new InMemoryTimeOrderedKeyValueBuffer();
+        timeDefinition = suppress.getTimeDefinition();
+        bufferFullStrategy = suppress.getBufferConfig().bufferFullStrategy();
+        shouldSuppressTombstones = suppress.shouldSuppressTombstones();
     }
 
     @Override
     public void init(final ProcessorContext context) {
         internalProcessorContext = (InternalProcessorContext) context;
+        this.keySerde = keySerde == null ? castSerde(context.keySerde()) : keySerde;
+        this.valueSerde = valueSerde == null ? FullChangeSerde.castAndWrap(context.valueSerde()) : valueSerde;
     }
 
     @Override
     public void process(final K key, final Change<V> value) {
-        if (suppress.getTimeToWaitForMoreEvents() == Duration.ZERO && definedRecordTime(key) <= internalProcessorContext.streamTime()) {
+        if (suppressDurationMillis == 0 && definedRecordTime(key) <= internalProcessorContext.streamTime()) {
             if (shouldForward(value)) {
                 internalProcessorContext.forward(key, value);
             } // else skip
         } else {
-            throw new NotImplementedException();
+            buffer(key, value);
+            enforceConstraints(internalProcessorContext.streamTime());
         }
     }
 
-    private boolean shouldForward(final Change<V> value) {
-        return !(value.newValue == null && suppress.suppressTombstones());
+    @SuppressWarnings("unchecked")
+    private <T> Serde<T> castSerde(final Serde<?> untyped) {
+        return (Serde<T>) untyped;
+    }
+
+    private void buffer(final K key, final Change<V> value) {
+        final long time = definedRecordTime(key);
+        final ProcessorRecordContext recordContext = internalProcessorContext.recordContext();
+
+        final Bytes serializedKey = Bytes.wrap(keySerde.serializer().serialize(null, key));
+        final byte[] serializedValue = valueSerde.serializer().serialize(null, value);
+
+        buffer.put(time, serializedKey, new ContextualRecord<>(serializedValue, recordContext));
     }
 
     private long definedRecordTime(final K key) {
-        return suppress.getTimeDefinition().time(internalProcessorContext, key);
+        return timeDefinition.time(internalProcessorContext, key);
     }
 
-    @Override
-    public void close() {
+    private void enforceConstraints(final long streamTime) {
+        final long expiryTime = streamTime - suppressDurationMillis;
+        if (overCapacity() || buffer.minTimestamp() <= expiryTime) {
+
+            drainExpiredRecords(expiryTime);
```

Comment:
Should we inline this method?

Also, I am wondering if we could/should call this unconditionally? If `overCapacity()` is true, we might or might not expire records here (same if called unconditionally). If `overCapacity()` is false, but `buffer.minTimestamp() <= expiryTime` is true, we would expire record (same if called unconditionally). If both are false, `drainExpriredRecords()` would not expire anything if called either, because it passed in the corresponding boolean predicate anyway?

Ie, I _think_ we can just remove the `if` condition and execute the `then` part always

Reply:
I wanted to save on setting up the iterator, but your comment made me realize we can and should do that with an initial `if (predicate.get())` inside `evictWhile`. I did this and removed the condition as you recommended.